tensornetwork/block_sparse/charge.py

```diff
@@ -264,6 +275,169 @@ def intersect(self, other, assume_unique=False, return_indices=False) -> Any:
+    if isinstance(target_charges, (np.integer, int)):
+      target_charges = np.asarray([target_charges], dtype=np.int16)
+    if target_charges.ndim == 1:
+      target_charges = np.expand_dims(target_charges, 0)
+    target_charges = np.asarray(target_charges, dtype=np.int16)
+    # find intersection of index charges and target charges
+    reduced_charges, label_to_unique, _ = intersect(
+        self.unique_charges, target_charges, axis=1, return_indices=True)
+    num_unique = len(label_to_unique)
+
+    # construct the map to the reduced charges
+    map_to_reduced = np.full(self.dim, fill_value=-1, dtype=np.int16)
+    map_to_reduced[label_to_unique] = np.arange(num_unique, dtype=np.int16)
+
+    # construct the map to the reduced charges
+    reduced_ind_labels = map_to_reduced[self.charge_labels]
+    reduced_locs = reduced_ind_labels >= 0
+    new_ind_labels = reduced_ind_labels[reduced_locs].astype(np.int16)
+    obj = self.__new__(type(self))
+    obj.__init__(reduced_charges, new_ind_labels, self.charge_types)
+
+    if return_locations:
+      return obj, strides * np.flatnonzero(reduced_locs).astype(np.uint32)
+    return obj
+
+  def __getitem__(self, n: Union[List[int], np.ndarray, int]) -> "BaseCharge":
+    """
+    Return the charge-element at position `n`, wrapped into a `BaseCharge`
+    object.
+    Args:
+      n: An integer or `np.ndarray`.
+    Returns:
+      BaseCharge: The charges at `n`.
+    """
+
+    if isinstance(n, (np.integer, int)):
+      n = np.asarray([n])
+    n = np.asarray(n)
+    obj = self.__new__(type(self))
+    labels = self.charge_labels[n]
+    unique_labels, new_labels = np.unique(labels, return_inverse=True)
+    if unique_labels.ndim == 0:
+      unique_labels = np.asarray(unique_labels)
+    unique_charges = self.unique_charges[:, unique_labels]
+    obj.__init__(unique_charges, new_labels, self.charge_types)
+    return obj
+
+  def isin(self, target_charges: Union[np.ndarray, "BaseCharge"]) -> np.ndarray:
+    """
+    See also np.isin. 
+    Returns an np.ndarray of `dtype=bool`, with `True` at all linear positions
+    where `self` is in `target_charges`, and `False` everywhere else.
+    Args:
+      target_charges: A `BaseCharge` object.
+    Returns:
+      np.ndarray: An array of boolean values.
+    """
+    if isinstance(target_charges, type(self)):
+      if not np.all([
+          a == b for a, b in zip(self.charge_types, target_charges.charge_types)
+      ]):
+        raise TypeError(
+            "isin only callable for equal charge types, found {} and {}".format(
+                self.charge_types, target_charges.charge_types))
+
+      targets = target_charges.unique_charges
+    else:
+      if target_charges.ndim == 1:
+        targets = np.expand_dims(np.unique(target_charges, axis=0), 0)
+      elif target_charges.ndim == 2:
+        targets = np.unique(target_charges, axis=1)
+      else:
+        raise ValueError("targets.ndim has to be 1 or 2, found {}".format(
```

Comment:
Please add tests that hit all of these raises. Tests on failure cases accounts for a majority of our missing coverage currently

Reply:
will do

Reply:
fixed